boot-speed/clouds/measure-cloud.py

```diff
@@ -190,6 +192,7 @@ def measure_instance(instance, datadir, reboots=1):
     for nboot in range(1, reboots+1):
         bootdir = "boot_" + str(nboot)
         instance.restart()
+        time.sleep(60)
```

Comment:
what is this for? doesn't restart have a wait flag?

Reply:
It doesn't, but it probably should:

https://git.launchpad.net/pycloudlib/tree/pycloudlib/ec2/instance.py#n109

Let's do like this if you agree: I'll add a `wait` flag to pycloudlib and update this PR to use it.

Reply:
ha! I remember why:

```
# This is a terrible hack, however it is not obviously clear
        # how to wait for a system to start to restart.
        time.sleep(30)
```

I'd like to avoid generically waiting 30 + 60 seconds and solve this in pycloudlib.

Reply:
This is how we solved this in cloud-test-framework:

```py
def _count_reboots_for_instance(instance):
    return len(
        instance.interactions.run_command('last reboot -F').splitlines())


class RebootInstanceTask(Task):

    def execute(self, test_metadata, instance):
        LOGGER.info('Rebooting instance "{}"...'.format(instance.instance_id))
        pre_reboot_count = _count_reboots_for_instance(instance)
        instance.interactions.trigger_reboot()
        down = False
        while not down:
            try:
                if _count_reboots_for_instance(instance) > pre_reboot_count:
                    break
            except (NetworkError, socket.error, EOFError):
                down = True
            time.sleep(1)
        instance.interactions.purge_ssh_connection()
        LOGGER.info('Instance down; proceeding.')
```

So, basically, determine the number of times the system has rebooted thus far before triggering the reboot. Then reboot and wait until either we detect that the number of reboots has increased (which means it completed the reboot between us attempting to determine if it was down), or if we detect that we can no longer connect to the instance we assume that it's in the process of rebooting.

(In c-t-f, tasks always wait for SSH to be available and cloud-init to complete, so just getting to "down" is sufficient there.)

Reply:
boto3 doesn't help:

> Requests a reboot of the specified instances. This operation is asynchronous; it only queues a request to reboot the specified instances. 
> 
> If an instance does not cleanly shut down within four minutes, Amazon EC2 performs a hard reboot.

One easy solution here is bumping the sleep time to 60s in pycloudlib. I observed 30s not being enough sometimes, and the default should be safe.

Another solution is adding a `sleep=30` parameter to the `restart()` method, to increase (or decrease) the sleep time without double-sleeping. *Looks* clean but it's a bit ad-hoc for EC2, as we may have better solutions for other clouds, and we want to keep the high-level pycloydlib interface uniform.

Reply:
I wrote the previous comment before reading @OddBloke's. Let's see what I can come up with.

Reply:
I dropped the commit adding the sleep and implemented a wait mechanism in pycloudlib:

https://code.launchpad.net/~legovini/pycloudlib/+git/pycloudlib/+merge/369113

Reply:
An improvement on yesterday's MP:

https://code.launchpad.net/~legovini/pycloudlib/+git/pycloudlib/+merge/369158

It handles the case where, in @OddBloke's code, this fails:
```
pre_reboot_count = _count_reboots_for_instance(instance)
```